feat: Реализована базовая функциональность трекера посылок

Shilan42 opened the pull request:
Добавлена работа с базой данных SQLite и реализованы CRUD-операции:
- Регистрация посылок с автогенерации трек-номера
- Получение списка посылок клиента
- Обновление статуса посылки
- Изменение адреса доставки
- Удаление посылок

tests: Добавлены интеграционные тесты

Реализован набор тестов для проверки:
- Корректности CRUD-операций
- Валидации данных
- Обработки ошибок
- Работоспособности всех функций

Все тесты успешно пройдены, функционал работает корректно.

Shilan42 committed:
feat: Реализована базовая функциональность трекера посылок

Добавлена работа с базой данных SQLite и реализованы CRUD-операции:
- Регистрация посылок с автогенерации трек-номера
- Получение списка посылок клиента
- Обновление статуса посылки
- Изменение адреса доставки
- Удаление посылок

tests: Добавлены интеграционные тесты

Реализован набор тестов для проверки:
- Корректности CRUD-операций
- Валидации данных
- Обработки ошибок
- Работоспособности всех функций

Все тесты успешно пройдены, функционал работает корректно.

Shilan42 committed:
Add GitHub Actions workflow for testing

Shilan42 committed:
удалил репозиторий с автотестами

ivannizh commented on parcel.go:
```diff
@@ -2,59 +2,163 @@ package main
+		}
+		// Добавляем посылку = структуру в слайс для хранения найденных посылок
+		res = append(res, p)
+	}
```
После цикла нужно проверить курсор на наличие ошибок

http://go-database-sql.org/retrieving.html

Shilan42 replied:
Привет!
Спасибо за обратную связь!

Добавил после цикла `for rows.Next()` проверку на наличие ошибок, возникших при итерации по всем строкам результата запроса:

`if err = rows.Err(); err != nil {`
		`return nil, fmt.Errorf("error iterating through rows while retrieving client's parcels %d: %w", client, err)`
	`}`

ivannizh commented on parcel.go:
```diff
@@ -2,59 +2,163 @@ package main
-
-	return nil
+	// Начинаем транзакцию для обеспечения целостности данных (решил попробовать поюзать транзакции)
+	tx, err := s.db.Begin()
```
В этой и следующей функции нужно упростить логику и делать все в одном запросе.

Нужно убрать первый запрос и добавить в `UPDATE`(`DELETE`) еще одно условие в секцию `WHERE`

ivannizh replied:
Тогда и сложности с транзакцией не нужны

Shilan42 replied:
Поправил функции `SetAddress` и `Delete` в файле parcel.go.

ivannizh commented on parcel_test.go:
```diff
@@ -28,94 +33,233 @@ func getTestParcel() Parcel {
-	// get
-	// получите только что добавленную посылку, убедитесь в отсутствии ошибки
-	// проверьте, что значения всех полей в полученном объекте совпадают со значениями полей в переменной parcel
+				assert.Equal(t, res.Address, parcel.Address, "address mismatch. Expected: %v, Actual: %v", parcel.Address, res.Address)
```
Функция `Equal` может сравнивать структуры сразу (а не по полям), надо только с полем `Number` разобраться.

Нужно переделать, так как при таком подходе при изменении структуры нужно будет менять все тесты

ivannizh commented on parcel_test.go:
```diff
@@ -28,94 +33,233 @@ func getTestParcel() Parcel {
+	tests := []struct {
+		name     string                                // Название тестового кейса
+		testFunc func(*testing.T, ParcelStore, Parcel) // Функция, реализующая логику теста
+	}{
```
Хорошие тест кейсы

ivannizh commented on parcel_test.go:
```diff
@@ -28,94 +33,233 @@ func getTestParcel() Parcel {
+				require.Error(t, err, "expected error when trying to retrieve deleted parcel with ID %d", parcel.Number)
+
+				originalErr := errors.Unwrap(err)
+				require.Equal(t, sql.ErrNoRows, originalErr, "expected specific sql.ErrNoRows error when searching for deleted parcel with ID %d", parcel.Number)
```
Для сравнивания ошибок нужно использовать специализированную функцию
```suggestion
	require.ErrorIs(t, originalErr, sql.ErrNoRows, "expected specific sql.ErrNoRows error when searching for deleted parcel with ID %d", parcel.Number)
```

ivannizh commented on parcel_test.go:
```diff
@@ -28,94 +33,233 @@ func getTestParcel() Parcel {
+		originalParcel, ok := parcelMap[parcel.Number]
+		require.True(t, ok, "parcel with ID %d not found in original data", parcel.Number)
+		// Проверка всех полей полученной посылки
+		assert.Equal(t, parcel.Address, originalParcel.Address, "address mismatch. Expected: %s, Actual: %s", originalParcel.Address, parcel.Address)
```
Аналогично нужно сравнивать всю структуру целиком.

ivannizh reviewed (commented):
Привет! Хорошая работа! Осталось исправить некоторые моменты и работа будет принята.

Shilan42 committed:
feat: Улучшение обработки ошибок и валидации операций с посылками

Основные изменения:
- Добавлена проверка ошибок итерации в parcel.GetByClient
- Улучшена обработка операций в parcel.SetAddress и parcel.Delete:
  * Добавлена проверка количества затронутых строк
  * Реализовано логирование при неудачных операциях
  * Усилена проверка статуса ParcelStatusRegistered
  * Улучшено форматирование сообщений об ошибках
- Обновлены тесты в parcel_test:
  * Усилена проверка корректности получения посылок
  * Добавлена проверка специфической ошибки sql.ErrNoRows
  * Улучшена валидация полей в TestGetByClient
  * Усилена проверка совпадения номера посылки

ivannizh reviewed (approved):
Отлично, принято!
